README.md
# Learn Design

An open, react-based set of experimental exercises for students to learn design components, focused on deliberately limiting tools, promoting games, and encouraging exploration.

- [Roadmap](https://github.com/ashleykolodziej/learn-design/projects/3)
- [Core Skills](https://github.com/ashleykolodziej/learn-design/projects/1)
- [Inspiration and Attribution](https://github.com/ashleykolodziej/learn-design/projects/2)

[![ko-fi](https://www.ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/Q5Q21MJMR)

## Get started

- Clone this directory
- cd to directory and `npm install`
- `npm run start` to begin editing
- `npm run build` to create a final build

CHANGELOG.md
# Unreleased

- Adds new kernable component via #19
- Switch to React and Sass via #19
- Cleanup and general project documentation

# 0.1.0

- Initial setup of repo

src/index.js
'use strict';

import "./styles.css";

const UI = {
  render: function(content, where = "beforeend") {
    const container = document.querySelector("#app");
    container.insertAdjacentHTML(where, content);
  }
};

UI.render( "<h1>Learn Design</h1>", "afterbegin" );

UI.render( `<div id="kerning" contenteditable="true">
	${stringToSpans( "Wave" )}
</div>`,

"afterbegin" );

var intendedDirection = null;

function stringToSpans( str ) {
	var newHTML = '';

	for ( var i = 0; i < str.length; i++ ) {
		newHTML += `<span>${str[i]}</span>`;
	}

	return newHTML;
}

function getSelected( intendedDirection ) {
	if ( window.getSelection ) {
		var sel = window.getSelection();

		if ( sel.type === "Range" ) {
			return sel.anchorNode.children;
		}

		if ( intendedDirection === "left" && sel.anchorNode ) {
			return sel.anchorNode.parentNode.previousSibling;
		} else {
			return sel.anchorNode.parentNode;
		}
	}

	return null;
}

function isIterable( obj ) {
	if ( obj === null ) {
		return false;
	}

	return typeof obj[Symbol.iterator] === 'function';
}

function resetIntendedDirection() {
	intendedDirection = null;
}

function getKerning( letter ) {
	var kerning = 0;

	if ( letter.style.letterSpacing !== "" ) {
		kerning = parseInt( letter.style.letterSpacing );
	}

	return kerning;
}

function adjustAmount( amount, direction ) {
	if ( direction === "left" ) {
		amount--;
	} else {
		amount++;
	}

	return amount;
}

function kernLetter( letter, direction ) {
	var amount = getKerning( letter );
	amount = adjustAmount( amount, direction );
	letter.style.letterSpacing = `${amount}px`;
}

function kernSelection( selection, direction ) {
	if ( selection.length ) {
		for ( var i = 0; i <= selection.length; i++ ) {
			kernLetter( selection[i], direction );
		}
	} else {
		kernLetter( selection, direction );
	}
}

function kernControls(e) {
	var pressed = e.keyCode,
		 altPressed = e.altKey,
		 metaPressed = e.metaKey,
		 selected = getSelected();

	if ( intendedDirection !== null ) {
		selected = getSelected( intendedDirection );
	}

	if ( altPressed ) {
		e.preventDefault();
	}

	// Prevent default for all keypresses except alt and arrows.
	switch( pressed ) {
		case 18: case 224:
			break;
		case 65:
			if ( !metaPressed ) {
				e.preventDefault();
			} else {
				selected = getSelected();
			}
			break;
		case 37:
			if ( altPressed ) {
				kernSelection( selected, "left" );
			} else {
				intendedDirection = "left";
			}

			break;

		case 39:
			if ( altPressed ) {
				kernSelection( selected, "right" );
			} else {
				intendedDirection = "right";
			}

			break;

		default:
			e.preventDefault();
	}
}

var kernArea = document.getElementById( `kerning` );

kernArea.addEventListener( 'keydown', kernControls );
kernArea.addEventListener( 'click', resetintendedDirection );


src/components/kern.js
import React from 'react';
import { stringToSpans } from './library';

class Kern extends React.Component {
	constructor (props) {
		super(props);
		this.state = {
			intendedDirection: null,
		};
	}

	resetIntendedDirection = () => {
		this.setState({ intendedDirection: null });
	}

	getKerning = ( letter ) => {
		var kerning = 0;

		if ( letter.style.letterSpacing !== "" ) {
			kerning = parseInt( letter.style.letterSpacing );
		}

		return kerning;
	}

	adjustAmount = ( amount, direction ) => {
		if ( direction === "left" ) {
			amount--;
		} else {
			amount++;
		}

		return amount;
	}

	getSelected = ( intendedDirection ) => {
		if ( window.getSelection ) {
			var sel = window.getSelection();

			if ( sel.type === "Range" ) {
				return sel.anchorNode.children;
			}

			if ( intendedDirection === "left" && sel.anchorNode ) {
				return sel.anchorNode.parentNode.previousSibling;
			} else {
				return sel.anchorNode.parentNode;
			}
		}

		return null;
	}

	kernLetter = ( letter, direction ) => {
		var amount = this.getKerning( letter );
		amount = this.adjustAmount( amount, direction );
		letter.style.letterSpacing = `${amount}px`;
	}

	kernSelection = ( selection, direction ) => {
		if ( selection.length ) {
			for ( var i = 0; i < selection.length; i++ ) {
				this.kernLetter( selection[i], direction );
			}
		} else {
			this.kernLetter( selection, direction );
		}
	}

	kernControls = (e) => {
		var pressed = e.keyCode,
			 altPressed = e.altKey,
			 metaPressed = e.metaKey,
			 selected = this.getSelected();

		if ( this.state.intendedDirection !== null ) {
			selected = this.getSelected( this.state.intendedDirection );
		}

		if ( altPressed ) {
			e.preventDefault();
		}

		// Prevent default for all keypresses except alt and arrows.
		switch( pressed ) {
			case 18: case 224:
				break;
			case 65:
				if ( !metaPressed ) {
					e.preventDefault();
				} else {
					selected = this.getSelected();
				}
				break;
			case 37:
				if ( altPressed ) {
					this.kernSelection( selected, "left" );
				} else {
					this.setState({ intendedDirection: "left" });
				}

				break;

			case 39:
				if ( altPressed ) {
					this.kernSelection( selected, "right" );
				} else {
					this.setState({ intendedDirection: "right" });
				}

				break;

			default:
				e.preventDefault();
		}
	}

	render() {
		let type = stringToSpans(this.props.word);
		return (
			<div id="kerning"
				contenteditable="true"
				onKeyDown={this.kernControls}
				onClick={this.resetIntendedDirection}
			>
				{type}
			</div>
		);
	}
}

export default Kern;

src/components/exercise/exercise.js
import React, { Component, Fragment } from 'react';
import Kernable from 'components/kernable/kernable';
import data from "./demo";
import './exercise.scss';

let exercise = data;
let component = exercise.component[0];

/**
* A set of explicitly supported components for exercises.
*
* Why?
*
* Because React needs the component name to be assigned to a capitalized
* variable in order to know to render it as a component and not HTML.
*
* See https://reactjs.org/docs/jsx-in-depth.html#choosing-the-type-at-runtime
*/

const Supported = {
	Kernable
};

class Exercise extends Component {
	constructor(props) {
		super(props);

		this.state = {
			isLoading: false
		};
	}

	/**
	* If there is an exercise name defined, load the exercise data.
	* Return an error if the exercise can't be found and show sample data instead.
	* Finally, hold off on rendering until we know for sure what's going on.
	*/

	async componentDidMount() {
		const exerciseName = this.props.name;

		if ( exerciseName ) {
			this.setState({ isLoading: true });
			const specifics = await import(`../../../data/exercises/${exerciseName}.json`).then(value => {
				exercise = value.default;
				component = exercise.component[0];
				this.setState({ isLoading: false });
			}, reason => {
			  this.setState({ isLoading: false });
			  console.error( "The requested exercise couldn't be found. Rendering the demo exercise." );
			});
		}
	}

	/**
	* Creates a React element based on an explicitly supported list of components
	* for this type of component. See Supported above.
	*/

	createElement = ( component, props = component.props ) => {
		return( React.createElement(
			Supported[component.type],
			props
		) );
	}

	render() {
		if ( this.state.isLoading ) return null;

		return (
			<Fragment>
				<section className="exercise">
					<h3 className="exercise-title">{exercise.title}</h3>
					<p className="exercise-directions">{exercise.directions}</p>
					{ this.createElement( component ) }
				</section>
			</Fragment>
		);
	}
}

export default Exercise;

src/components/kernable/kernable.js
import React from 'react';
import { stringToSpans } from 'components/library';
import './kernable.scss';

/**
* "Intent" only happens when you hit an arrow, so when you
* first encounter a kernable word, you haven't shown intent yet.
*/

const initialState = {
	intendedDirection: null,
};

class Kernable extends React.Component {
	/**
	* Build a kernable word from settings, and help our event
	* listeners understand the context of "this".
	*/
	constructor (props) {
		super(props);
		this.state = {
			...initialState,
		};

		this.kern = this.kern.bind(this);
		this.resetState = this.resetState.bind(this);
	}

	/**
	* Resets the state to null.
	*/
	resetState() {
		this.setState({
			intendedDirection: null,
		});
	}

	/**
	* Get the letter or letters to kern, based on the user selection
	* and with assumptions about intended direction based on previous
	* arrow press. ☕
	*/
	getSelected( intendedDirection ) {
		if ( window.getSelection ) {
			var sel = window.getSelection();

			if ( sel.type === "Range" ) {
				return sel.anchorNode.children;
			}

			if ( intendedDirection === "left" && sel.anchorNode ) {
				return sel.anchorNode.parentNode.previousSibling;
			} else {
				return sel.anchorNode.parentNode;
			}
		}

		return null;
	}

	/**
	* Get the existing kerning value, if it exists.
	*/
	getKerning( letter ) {
		var kerning = 0;

		if ( letter.style.letterSpacing !== "" ) {
			kerning = parseInt( letter.style.letterSpacing );
		}

		return kerning;
	}

	/**
	* Adjust kerning value up or down based on pressed arrow direction.
	*/
	adjustAmount( amount, direction ) {
		if ( direction === "left" ) {
			amount--;
		} else {
			amount++;
		}

		return amount;
	}

	/**
	* Kern a single letter based on user input.
	*/
	kernLetter( letter, direction ) {
		var amount = this.getKerning( letter );
		amount = this.adjustAmount( amount, direction );
		letter.style.letterSpacing = `${amount}px`;
	}

	/**
	* Kern the current selection based on user input.
	* Supports both single and multiple letter selections.
	*/
	kernSelection( selection, direction ) {
		if ( selection.length ) {
			for ( var i = 0; i < selection.length; i++ ) {
				this.kernLetter( selection[i], direction );
			}
		} else {
			this.kernLetter( selection, direction );
		}
	}

	/**
	* Kern, baby kern!
	* Disallows all keypresses except alt, meta (windows/mac key), and arrows.
	* Sets the intended direction or requests kerning based on keypress values.
	*/
	kern(e) {
		var pressed = e.keyCode,
			 altPressed = e.altKey,
			 metaPressed = e.metaKey,
			 selected = this.getSelected();

		if ( this.state.intendedDirection !== null ) {
			selected = this.getSelected( this.state.intendedDirection );
		}

		if ( altPressed ) {
			e.preventDefault();
		}

		// Prevent default for all keypresses except alt, meta, and arrows.
		switch( pressed ) {
			case 18: case 224:
				break;
			case 65:
				if ( !metaPressed ) {
					e.preventDefault();
				} else {
					selected = this.getSelected();
				}
				break;
			case 37:
				if ( altPressed ) {
					this.kernSelection( selected, "left" );
				} else {
					this.setState({ intendedDirection: "left" });
				}

				break;

			case 39:
				if ( altPressed ) {
					this.kernSelection( selected, "right" );
				} else {
					this.setState({ intendedDirection: "right" });
				}

				break;

			default:
				e.preventDefault();
		}
	}

	/**
	* Go time!
	*
	* contentEditable is used to allow cursor and selection only.
	* This content isn't actually editable because we disable all key
	* default behaviors besides alt, meta, and arrows.
	*/
	render() {
		return (
			<div className="kernable"
				contentEditable="true"
				suppressContentEditableWarning={true}
				onKeyDown={this.kern}
				onClick={this.resetState}>
				{stringToSpans(this.props.word)}
			</div>
		);
	}
}

export default Kernable;